device/common/include/traccc/seeding/device/impl/update_triplet_weights.ipp

```diff
@@ -0,0 +1,124 @@
+
+    // Current work item
+    triplet this_triplet = triplets_per_bin[ps_idx.second];
+
+    const sp_location& spT_idx = this_triplet.sp3;
+
+    const traccc::internal_spacepoint<traccc::spacepoint> current_spT =
+        sp_grid.bin(spT_idx.bin_idx)[spT_idx.sp_idx];
+
+    const scalar currentTop_r = current_spT.radius();
+
+    // if two compatible seeds with high distance in r are found, compatible
+    // seeds span 5 layers
+    // -> very good seed
+    const scalar lowerLimitCurv =
+        this_triplet.curvature - filter_config.deltaInvHelixDiameter;
+    const scalar upperLimitCurv =
+        this_triplet.curvature + filter_config.deltaInvHelixDiameter;
+    std::size_t num_compat_seedR = 0;
+
+    static constexpr std::size_t MAX_COMPAT_SEED = 5;
+    assert(MAX_COMPAT_SEED >= filter_config.compatSeedLimit);
+    scalar compat_seedR[MAX_COMPAT_SEED];
+
+    // iterate over triplets
+    for (auto tr_it = triplets_per_bin.begin(); tr_it != triplets_per_bin.end();
+         tr_it++) {
+        // only use other triplets which share same midBot doublet
+        if (this_triplet == *tr_it || !(this_triplet.sp1 == tr_it->sp1) ||
+            !(this_triplet.sp2 == tr_it->sp2)) {
+            continue;
+        }
+
+        const triplet& other_triplet = *tr_it;
+        const sp_location other_spT_idx = other_triplet.sp3;
+        const traccc::internal_spacepoint<traccc::spacepoint> other_spT =
+            sp_grid.bin(other_spT_idx.bin_idx)[other_spT_idx.sp_idx];
+
+        // compared top SP should have at least deltaRMin distance
+        const scalar otherTop_r = other_spT.radius();
+        const scalar deltaR = currentTop_r - otherTop_r;
+        if (std::abs(deltaR) < filter_config.deltaRMin) {
+            continue;
+        }
+
+        // curvature difference within limits?
+        // TODO: how much slower than sorting all vectors by curvature
+        // and breaking out of loop? i.e. is vector size large (e.g. in
+        // jets?)
+        if (other_triplet.curvature < lowerLimitCurv) {
+            continue;
+        }
+        if (other_triplet.curvature > upperLimitCurv) {
+            continue;
+        }
+
+        bool newCompSeed = true;
+
+        for (std::size_t i_s = 0; i_s < num_compat_seedR; ++i_s) {
+            const scalar previousDiameter = compat_seedR[i_s];
+
+            // original ATLAS code uses higher min distance for 2nd found
+            // compatible seed (20mm instead of 5mm) add new compatible seed
+            // only if distance larger than rmin to all other compatible
+            // seeds
+            if (std::abs(previousDiameter - otherTop_r) <
+                filter_config.deltaRMin) {
+                newCompSeed = false;
+                break;
+            }
+        }
+
+        if (newCompSeed) {
+            compat_seedR[num_compat_seedR] = otherTop_r;
+            triplets_per_bin[ps_idx.second].weight +=
+                filter_config.compatSeedWeight;
```

Comment:
Ahh... This is not what I had in mind...

Let's use `this_triplet` in this line. Let's just update the weight in the thread-specific variable during the for-loop.

Then, at the very end of the function, after the for-loop, let's just write to the global memory once.

```c++
triplets_per_bin[ps_idx.second].weight = this_triplet.weight;
```